include/gsl/joining_thread

```diff
@@ -0,0 +1,78 @@
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright (c) 2015 Microsoft Corporation. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#ifndef GSL_JOINING_THREAD_H
+#define GSL_JOINING_THREAD_H
+
+#include <memory>
+#include <thread>
+#include <utility>
+
+namespace gsl
+{
+/**
+ * RAII wrapper for std::thread.
+ * It's guaranteed that joining_thread will join when destructed.
+ */
+class joining_thread
+{
+    using JoiningThreadUPtr = std::unique_ptr<std::thread, std::function<void(std::thread*)>>;
+    template <typename... Args>
+    JoiningThreadUPtr create_thread_with_joining_deleter(Args&&... args)
+    {
+        return {new std::thread(std::forward<Args>(args)...), [](auto* th) { th->join(); }};
```

Comment:
I see a `new` with no corresponding `delete`: Where is this `std::thread` instance deleted ? the deleter you provide to `std::unique_ptr` only joins without deleting

Reply:
you're right it's an oversight. I've removed unique_ptr completely.